### PR TITLE
debian: add snapd.postrm that purges

### PR DESCRIPTION
**Pull request by mvo5**
`apt purge snapd` will purge all of the full snapd state and all snaps from the system.

### CODE REVIEW

File: debian/snapd.postrm [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+set -e
+
+if [ "$1" = "purge" ]; then
+   # FIXME: should we try to remove all snaps, mount points, services,
+   #        mount units that got created as well?
+   rm -f /var/lib/snapd/state.json
+fi
```
Comment (pedronis) [inline debian/snapd.postrm]: missing new line

File: debian/snapd.postrm [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+set -e
+
+if [ "$1" = "purge" ]; then
+   # FIXME: should we try to remove all snaps, mount points, services,
+   #        mount units that got created as well?
+   rm -f /var/lib/snapd/state.json
```
Comment (pedronis) [inline debian/snapd.postrm]: can we echo some kind of warnings? maybe listing the units we likely are orphaning?

**Comment by pedronis** [comment]
as seen in #1392 this indeed helps not confusing reinstalls downgrading the package

**Comment by mvo5** [comment]
retest this please

File: debian/snapd.postrm [hunk 3]
```diff
@@ -3,7 +3,19 @@
...
-   #        mount units that got created as well?
-   rm -f /var/lib/snapd/state.json
-fi
\ No newline at end of file
+
+    for snap in /var/lib/snapd/snaps/*.snap; do
+        base="$(basename $snap)"
+        # mount units
+        mangledbase="$(echo $base|tr _ -)"
+        mountunit="$(systemd-escape -p "${mangedbase%%.snap}".mount)"
```
Comment (niemeyer) [inline debian/snapd.postrm]: Typo in var name.

Related but off-topic note: can we change these filenames to not escape?  Our current conventions for snap and app name make them safe.

File: debian/snapd.postrm [hunk 4]
```diff
@@ -3,7 +3,19 @@
...
+        rm -f /etc/systemd/system/snap-$mountunit
+        rm -f /etc/systemd/system/multi-user.target.wants/snap-$mountunit
+
+        # FIXME: umount snaps
```
Comment (niemeyer) [inline debian/snapd.postrm]: Yeah, I think we want to do that before we remove these unit files. Seems strange to remove and leave the content behind.

Note we'll need to:

1. Stop service units
1. Stop mount units
1. Remove service and mount units
1. Remove all content

The tests/lib/reset.sh actually holds exactly this logic, except your method of finding out mount units is nicer above.

File: debian/snapd.postrm [hunk 5]
```diff
@@ -2,20 +2,32 @@
...
+    mounts="$(systemctl list-unit-files | grep '^snap[-.].*\.mount' | cut -f1 -d ' ')"
+    services="$(systemctl list-unit-files | grep '^snap[-.].*\.service' | cut -f1 -d ' ')"
+    for unit in $services $mounts; do
+        echo "Stopping $unit"
+        systemctl_stop $unit
 
-        # FIXME: umount snaps
-        # FIXME: remove /snap
-        # FIXME: remove /var/snap
+        echo "Removing $mountunit"
```
Comment (pedronis) [inline debian/snapd.postrm]: $unit here?

File: tests/lib/reset.sh [hunk 6]
```diff
@@ -2,18 +2,8 @@
...
-    systemctl stop $unit || true
-done
-
-rm -f /tmp/ubuntu-core*
-rm -rf /snap/*
-rm -rf /var/lib/snapd/*
-rm -f /etc/systemd/system/snap[-.]*.{mount,service}
-rm -f /etc/systemd/system/multi-user.target.wants/snap[-.]*.{mount,service}
+# purge all state
+$PROJECT_PATH/debian/snapd.postrm purge
```
Comment (pedronis) [inline tests/lib/reset.sh]: I think the style is to use SPREAD_PATH here (PROJECT_PATH is a spread.yaml thing)

File: tests/lib/reset.sh [hunk 7]
```diff
@@ -3,17 +3,11 @@
...
 
 systemctl stop snapd.service snapd.socket
-mounts="$(systemctl list-unit-files | grep '^snap[-.].*\.mount' | cut -f1 -d ' ')"
-services="$(systemctl list-unit-files | grep '^snap[-.].*\.service' | cut -f1 -d ' ')"
-for unit in $services $mounts; do
-    systemctl stop $unit || true
-done
+
+# purge all state
+sh $PROJECT_PATH/debian/snapd.postrm purge
```
Comment (pedronis) [inline tests/lib/reset.sh]: I think the style is to use SPREAD_PATH here (PROJECT_PATH is a spread.yaml detail)

File: cmd/snap/cmd_run.go [hunk 8]
```diff
@@ -76,6 +77,10 @@ func (x *cmdRun) Execute(args []string) error {
 		return snapRunHook(x.Positional.SnapApp, x.Hook, x.Revision)
 	}
 
+	if cmd.Run.DebugShell {
+		x.Command += ":debug-shell"
```
Comment (pedronis) [inline cmd/snap/cmd_run.go]: this looks  like picked up from some other work

File: debian/snapd.postrm [hunk 9]
```diff
@@ -2,8 +2,32 @@
...
+        echo "Removing $mount"
+        rm -f /etc/systemd/system/$unit
+        rm -f /etc/systemd/system/multi-user.target.wants/$unit
+    done
+
+    echo "Removing snap directories (/snap)"
+    rm -rf /snap
+
+    echo "Removing snap data (/var/snap)"
+    rm -rf /var/snap
```
Comment (zyga) [inline debian/snapd.postrm]: I really don't think we should do this. This is like removing your database because you happen to remove postgresql. This is not something anyone would expect IMHO

File: tests/lib/reset.sh (hunk 7 above)
Comment (zyga) [inline tests/lib/reset.sh]: Thanks for doing this, I should be able to reuse this in packaging for various other distributions

File: cmd/snap/cmd_run.go [hunk 10]
```diff
@@ -45,9 +45,10 @@ type cmdRun struct {
 		SnapApp string `positional-arg-name:"<app name>" description:"the snap (e.g. hello-world) or application to run (e.g. hello-world.env)"`
 	} `positional-args:"yes" required:"yes"`
 
-	Command  string `long:"command" description:"alternative command to run" hidden:"yes"`
-	Hook     string `long:"hook" description:"hook to run" hidden:"yes"`
-	Revision string `short:"r" description:"use a specific snap revision when running hook" hidden:"yes"`
+	Command    string `long:"command" description:"alternative command to run" hidden:"yes"`
+	Hook       string `long:"hook" description:"hook to run" hidden:"yes"`
+	Revision   string `short:"r" description:"use a specific snap revision when running hook" hidden:"yes"`
+	DebugShell string `long:"debug-shell" description:"drop into a shell with the same confinment as the snap"`
```
Comment (zyga) [inline cmd/snap/cmd_run.go]: Seems like unrelated change

File: debian/snapd.postrm (hunk 9 above)
Comment (niemeyer) [inline debian/snapd.postrm]: If you purge postgresql the database actually goes with it.

File: tests/lib/reset.sh [hunk 11]
```diff
@@ -3,17 +3,11 @@
...
 
 systemctl stop snapd.service snapd.socket
-mounts="$(systemctl list-unit-files | grep '^snap[-.].*\.mount' | cut -f1 -d ' ')"
-services="$(systemctl list-unit-files | grep '^snap[-.].*\.service' | cut -f1 -d ' ')"
-for unit in $services $mounts; do
-    systemctl stop $unit || true
-done
+
+# purge all state
+sh ${SPREAD_PATH}/debian/snapd.postrm purge
```
Comment (zyga) [inline tests/lib/reset.sh]: Neat :-)

**Comment by pedronis** [comment]
LGTM once the tests pass

**Comment by zyga** [comment]
+1

**Comment by niemeyer** [comment]
The test breakages seem real and local to the branch.
